include/Reader/reader.h

```diff
@@ -1020,19 +1019,92 @@ EHRegion *getFinallyRegion(EHRegion *TryRegion);
 ///
 ///@{
 
-/// \brief Obtain a list of the successor edges of a FlowGraphNode
+/// \brief Class implementing an iterable list of flow graph edges.
+///
+/// Abstract class that serves as a common base for the client-supplied
+/// predecessor and successor iterator implementations.
+class FlowGraphEdgeIteratorImpl {
+public:
+  /// \brief Check if the iterator is at end.
+  ///
+  /// \returns True if there are no more edges to iterate.
+  virtual bool isEnd() = 0;
+
+  /// \brief Advance the iterato
+  virtual void moveNext() = 0;
+
+  /// \brief Obtain the sink for the current edge.
+  /// \returns The sink (aka target or destination) node of the current edge.
+  virtual FlowGraphNode *getSink() = 0;
+
+  /// \brief Obtain the source for the current edge.
+  /// \return The source (aka From) node of the current ed
+  virtual FlowGraphNode *getSource() = 0;
+};
+
+/// \brief Class representing an iterable list of flow graph edges.
+///
+/// An iterator for flow graph edges that does not assume there is an actual
+/// object representing the edge. At construction time specify whether the
+/// iterator can traverse the successor or predecessor edges of a block.
+class FlowGraphEdgeIterator {
+public:
+  /// \brief Construct an iterator.
+  ///
+  /// \param Block         The basic block to use as edge source or sink.
+  /// \param IsSuccessor   If true, iterate the successor edges of the block.
+  FlowGraphEdgeIterator(FlowGraphNode *Block, bool IsSuccessor);
```

Comment:
It was my impression that those `fg*`  functions were done that way simply because they were (initially) stubbed out and so didn't need to be methods.  I've already had to move a few to be methods in the EH branch where I needed the reader object to give them real implementations.  I presume that if we wanted to support "multi-client" we'd just make them all virtual methods on `ReaderBase`.  And I guess the equivalent here would be to give `ReaderBase` a virtual factory method for FlowGraphEdgeIterators and make `Impl` settable.